### PR TITLE
tls: load certs from Windows system store

### LINKED ISSUE #61: Certificate verification fails on Windows native build
The http3-client fails certificate verification when built natively, e.g.

```
$ http3-client https://cloudflare-quic.com
snip...
[2019-06-15T11:43:24.931343300Z TRACE quiche::tls] error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
[2019-06-15T11:43:24.937788700Z ERROR http3_client] 88dab5f9b993280072028d5b4d34d30a4e3c recv failed: TlsFail
```

The cert for cloudflare-quic.com is trusted by my machine. So I suspect this is something related to building quiche in a way that properly imports my system trust store.

### PR DESCRIPTION
**Pull request by LPardue**
Rough first stab at this, PTAL. Fixes #61 .

Since this problem and its solution are windows specific, there is liberal peppering of `#[cfg(windows)]`.  I'm keen to see if this actually works as it appears on paper.

Happy to hear more elegant ways to do this. Tidy up code is not great, windows API example code is a bit rubbish.

### CODE REVIEW

**Comment by LPardue** [comment]
Ha! Clippy build failures because of unused functions...

**Comment by LPardue** [comment]
> it really shouldn't be exposed as public API, and certainly not one that is only available on Windows. The loading should happen in the same place as other trust stores are configured:
https://github.com/cloudflare/quiche/blob/master/src/tls.rs#L128

That sounds way more logical, thnx.

**Comment by junhochoi** [comment]
I have almost zero knowledge on Windows so it's hard to comment but I agree that user don't need to call anything additional API only for Windows.

This PR looks loading system certs, meaning in unix-like platforms BoringSSL does it automatically but not in Windows?

**Comment by LPardue** [comment]
> This PR looks loading system certs, meaning in unix-like platforms BoringSSL does it automatically but not in Windows?

AFAICT, `SSL_CTX_set_default_verify_paths()` work on Linux and MacOS but not Windows (google of the function and windows finds loads of results). I missed that we use this function, so it is the natural place to add platform-specific CA loading.

**Comment by LPardue** [comment]
Now the change only really affect tls.rs; this is much clearer.

File: src/tls.rs [hunk 1]
```diff
@@ -152,6 +179,47 @@ impl Context {
         })
     }
 
+    #[cfg(windows)]
+    fn load_ca_certs(&mut self) -> Result<()> {
```
Comment (ghedo) [inline src/tls.rs]: This method name is still pretty generic for being Windows-specific.

How about you define a separate implementation of the method for ``#[cfg(not(windows))]` that only does the `SSL_CTX_set_default_verify_paths()` call, so then in `new()` you'd just call `load_ca_certs()` unconditionally.

File: src/tls.rs [hunk 2]
```diff
@@ -679,6 +747,15 @@ extern {
 
     fn SSL_CTX_set_default_verify_paths(ctx: *mut SSL_CTX) -> c_int;
 
+    #[cfg(windows)]
+    fn SSL_CTX_get_cert_store(ctx: *mut SSL_CTX) -> *mut X509_STORE;
+    #[cfg(windows)]
+    fn X509_STORE_add_cert(ctx: *mut X509_STORE, x: *mut X509) -> c_int;
```
Comment (ghedo) [inline src/tls.rs]: Nitpick, but the functions in this block are organized by type, so you should move this and the `X509` functions to the end of the block and split them with `// X509_STORE` and `// X509` comments.

File: src/tls.rs [hunk 3]
```diff
@@ -25,6 +25,18 @@
 // NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 // SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+#[cfg(windows)]
+extern crate winapi;
+#[cfg(windows)]
+use winapi::um::wincrypt::{
+    CertCloseStore,
```
Comment (ghedo) [inline src/tls.rs]: Nit: generally we don't group imports like this, we simply expand all the `use`s.

File: src/tls.rs [hunk 4]
```diff
@@ -25,6 +25,18 @@
 // NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 // SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+#[cfg(windows)]
+extern crate winapi;
```
Comment (ghedo) [inline src/tls.rs]: Do you really need this?

Comment (ghedo) [reply src/tls.rs]: As in, you generally don't need to explicitly import crates that are defined in `Cargo.toml`, though I don't know if this is special in some way.

File: src/tls.rs (hunk 2 above)
Comment (LPardue) [reply src/tls.rs]: bloody d2i_, I folded that into ``// X509`` but leem know if you want a separate block

File: src/tls.rs (hunk 3 above)
Comment (LPardue) [reply src/tls.rs]: removed all imports/usage

File: src/tls.rs (hunk 1 above)
Comment (LPardue) [reply src/tls.rs]: I've done this but don't have a non-windows machine handy, so may find it breaks

File: src/tls.rs [hunk 5]
```diff
@@ -152,6 +164,60 @@ impl Context {
         })
     }
 
+    #[cfg(not(windows))]
+    fn load_ca_certs(&mut self) -> Result<()> {
+        unsafe { map_result(SSL_CTX_set_default_verify_paths(ctx_raw)) }
```
Comment (LPardue) [inline src/tls.rs]: self.as_ptr()
```suggestion
        unsafe { map_result(SSL_CTX_set_default_verify_paths(self.as_ptr())) }
```

**Pull request merged**